src/goto-instrument/contracts/contracts.cpp

```diff
@@ -779,28 +810,40 @@ bool code_contractst::check_frame_conditions_function(const irep_idt &function)
       ns.lookup(parameter.get_identifier()).symbol_expr());
   }
 
+  auto instruction_it = function_obj->second.body.instructions.begin();
+  for(const auto &car : assigns.get_write_set())
+  {
+    auto snapshot_instructions = car.generate_snapshot_instructions();
+    insert_before_swap_and_advance(
+      function_obj->second.body, instruction_it, snapshot_instructions);
+  };
+
   // Insert aliasing assertions
   check_frame_conditions(
-    function_obj->first, function_obj->second.body, assigns);
+    function_obj->first, function_obj->second.body, instruction_it, assigns);
 
   return false;
 }
 
 void code_contractst::check_frame_conditions(
   const irep_idt &function,
   goto_programt &body,
+  goto_programt::targett &instruction_it,
   assigns_clauset &assigns)
 {
   goto_function_inline(goto_functions, function, ns, log.get_message_handler());
 
-  for(auto instruction_it = body.instructions.begin();
-      instruction_it != body.instructions.end();
-      ++instruction_it)
+  for(; instruction_it != body.instructions.end(); ++instruction_it)
   {
     if(instruction_it->is_decl())
     {
-      // Local variables are always freely assignable
-      assigns.add_to_write_set(instruction_it->get_decl().symbol());
+      const auto car =
+        assigns.add_to_write_set(instruction_it->get_decl().symbol());
+      auto snapshot_instructions = car->generate_snapshot_instructions();
+      instruction_it++;
+      insert_before_swap_and_advance(
+        body, instruction_it, snapshot_instructions);
+      instruction_it--;
```

Comment:
Why do we need to modify `instruction_it` like this here?

Reply:
Added a comment here.